text/0000-embroider-v2-package-format.md

```diff
@@ -0,0 +1,817 @@
+
+- it makes Fastboot builds simpler because we can guard the fastboot-only and browser-only code with the macro system. In development, we can run a single build that leaves both branches in and evaluates the macros at runtime.
+
+The macros package (`@ember/macros` as proposed, `@embroider/macros` as implemented) will work in both regular ember-cli and in Embroider. And it will work in both V1 and V2 packages.
+
+## Peer Dependencies
+
+V2 packages can only resolve their **allowed dependencies**. This is fundamental rule that we can't break if we want the broadest compatibility with NPM and future compatibility with other strict systems such as [Yarn PnP](https://github.com/yarnpkg/rfcs/pull/101). Node often allows you to resolve things that are not **allowed dependencies** due to hoisting optimizations. But this is not safe or guaranteed, so we forbid relying on it.
+
+This means that many things addons will try to access from their surrounding environment will need to be listed as `peerDependencies`. For example, addons that want to import `ember-data` should list `ember-data` as a `peerDependency`, so the app can control the `ember-data` version and the addon is guaranteed to resolve the same copy.
+
+This also applies recursively -- if your addon wants to use an addon that needs `ember-data`, your addon should also list `ember-data` as a `peerDependency`. The clearest documented description of how recursive peerDependencies should work is in the [Yarn PnP Formal Guarantees](https://github.com/yarnpkg/rfcs/blob/master/accepted/0000-plug-an-play.md#c-formal-plugnplay-guarantees).
+
+`ember-source` provides many "virtual" packages like `@ember/component`. If they were real packages, they would be `peerDependencies`, but having non-real packages in package.json is likely to result in errors. Pedantically, they can be listed in **externals** instead. In practice, they are a well-known set that we can always handle correctly automatically, so it's not very important whether an addon includes them in **externals**.
+
+### Optional Peer Dependencies
+
+Some addons optionally use another addon if it happens to be available in the app. In order to resolve such a dependency, we really need **Optional Peer Dependencies**.
+
+NPM doesn't have a concept of optional peer dependencies. It has "optional dependencies", but they are something different and pretty useless.
+
+Yarn did an [RFC for optional peer dependency support](https://github.com/yarnpkg/rfcs/blob/master/accepted/0000-optional-peer-dependencies.md). It is basically compatible with NPM, with the only caveat being that if you use NPM you may see a spurious warning at install time. As non-actionable peerDependency warnings are rife throughout the NPM ecosystem this doesn't seem like a big deal.
```

Comment:
That RFC doesn't mention how one would consume such an addon. What happens if you try to `import` from a missing dep?

Reply:
import from a missing dep would be a static build error.

To actually import things out of an optional dependency you'd need to guard it with the macro system:

```js
if (macroCondition(dependencySatisfies("the-optional-thing", "*"))) {
  let thing = importSync("the-optional-thing");
  doSomethingWith(thing);
}
```